src/crimson/os/cyan_store.cc

```diff
@@ -4,6 +4,228 @@
+    throw std::runtime_error("mkfs");
+  }
+
+  string fn = path + "/collections";
+  bufferlist bl;
+  set<coll_t> collections;
+  encode(collections, bl);
+  r = bl.write_file(fn.c_str());
+  if (r < 0)
+    throw std::runtime_error("write_file");
+
+  write_meta("type", "memstore");
+  return seastar::now();
+}
+
+CyanStore::CollectionRef CyanStore::create_new_collection(const coll_t& cid)
+{
+  auto c = new Collection{cid};
+  return new_coll_map[cid] = c;
+}
+
+CyanStore::CollectionRef CyanStore::open_collection(const coll_t& cid)
+{
+  auto cp = coll_map.find(cid);
+  if (cp == coll_map.end())
+    return {};
+  return cp->second;
+}
+
+std::vector<coll_t> CyanStore::list_collections()
+{
+  std::vector<coll_t> collections;
+  for (auto& coll : coll_map) {
+    collections.push_back(coll.first);
+  }
+  return collections;
+}
+
+seastar::future<bufferlist> CyanStore::read(CollectionRef c,
+                                            const ghobject_t& oid,
+                                            uint64_t offset,
+                                            size_t len,
+                                            uint32_t op_flags)
+{
+  logger().info("{} {} {} {}~{}",
+                __func__, c->cid, oid, offset, len);
+  if (!c->exists) {
+    throw std::runtime_error(fmt::format("collection does not exist: {}", c->cid));
+  }
+  ObjectRef o = c->get_object(oid);
+  if (!o) {
+    throw std::runtime_error(fmt::format("object does not exist: {}", oid));
+  }
+  if (offset >= o->get_size())
+    return seastar::make_ready_future<bufferlist>();
+  size_t l = len;
+  if (l == 0 && offset == 0)  // note: len == 0 means read the entire object
+    l = o->get_size();
+  else if (offset + l > o->get_size())
+    l = o->get_size() - offset;
+  bufferlist bl;
+  if (int r = o->read(offset, l, bl); r < 0) {
+    throw std::runtime_error("read");
+  }
+  return seastar::make_ready_future<bufferlist>(std::move(bl));
+}
+
+seastar::future<> CyanStore::do_transaction(CollectionRef ch,
+                                            Transaction&& t)
+{
+  auto i = t.begin();
+  while (i.have_op()) {
+    Transaction::Op* op = i.decode_op();
+    int r = 0;
+    switch (op->op) {
+    case Transaction::OP_NOP:
+      break;
+    case Transaction::OP_WRITE:
+      {
+        coll_t cid = i.get_cid(op->cid);
+        ghobject_t oid = i.get_oid(op->oid);
+        uint64_t off = op->off;
+        uint64_t len = op->len;
+        uint32_t fadvise_flags = i.get_fadvise_flags();
+        bufferlist bl;
+        i.decode_bl(bl);
+        r = _write(cid, oid, off, len, bl, fadvise_flags);
+      }
+      break;
+    case Transaction::OP_MKCOLL:
+      {
+        coll_t cid = i.get_cid(op->cid);
+        r = _create_collection(cid, op->split_bits);
+      }
+      break;
+    default:
+      logger().error("bad op {}", static_cast<unsigned>(op->op));
+      abort();
+    }
+    if (r < 0) {
+      abort();
+    }
+  }
+  return seastar::now();
+}
+
```

Comment:
do we need return future for write, since for other store write will be blocked.

Reply:
if it's slow and is detected by the stall detecter, we can just rewrite it. but so far it works fine.